ops/pebble.py

```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+
+    def _state_parse_header(self):
+        next_state = None
+        try:
+            headers, content = self._buffer.split(b'\r\n\r\n', 1)
+        except ValueError:
+            # Not enough data; stop for now
+            self._update_initial_state(self._state_parse_header)
+            next_state = self._state_stop
+        else:
+            # Rely on the stdlib for parsing the headers, since this can be done purely
+            # in memory without any realistic worry of memory concerns.
+            parser = email.parser.BytesFeedParser()
+            parser.feed(headers)
+            message = parser.close()
+
+            content_disposition = message.get_content_disposition()
+            if content_disposition != 'form-data':
+                raise RuntimeError(
+                    'unexpected content disposition: {}'.format(content_disposition))
+
+            name = message.get_param('name', header='content-disposition')
+            if name == 'response':
+                next_state = self._state_parse_response_part
+            elif name == 'files':
+                filename = message.get_filename()
+                if filename is None:
+                    raise NotImplementedError('multipart "files" part missing filename')
+                self._prepare_tempfile(filename)
+                next_state = self._state_parse_files_part
+            else:
+                raise RuntimeError(
+                    'unexpected name in content-disposition header: {}'.format(name))
+
+            # Consume the header; now we're ready to parse the part body
+            # (either "request" or "files")
+            self._buffer = content
+        return next_state
+
+    def _state_parse_response_part(self):
+        # Response is handled purely in memory; wait on parsing until we've received
+        # the full part.
+        next_state = None
+        match = self._boundary_rgx.search(self._buffer)
+        if match:
+            span = match.span()
+            # Content is terminated by \r\n<boundary>; adjust for the \r\n
+            end_of_content = span[0] - 2
+            content = self._buffer[:end_of_content]
+
+            self._response = json.loads(content.decode())
+
+            self._buffer = self._buffer[span[0]:]
+            next_state = self._state_detect_boundary
+        else:
+            self._update_initial_state(self._state_parse_response_part)
+            next_state = self._state_stop
+        return next_state
+
+    def _state_parse_files_part(self):
+        # Files can be huge; we shouldn't keep them in memory; rather we should
+        # stream them to files.  For now, using tempfiles.
+        next_state = None
+        match = self._boundary_rgx.search(self._buffer)
+        if match:
+            span = match.span()
+            # Content is terminated by \r\n<boundary>; adjust for the \r\n
+            end_of_content = span[0] - 2
+            content = self._buffer[:end_of_content]
```

Comment:
I've got a chance to review all of your comments.  I'm not sure how soon time will allow me to address them in code, but I'd like to at least reply more completely.  I'll keep it brief though.

* Understood re: error/log strings and lower- vs. title-casing.  I'll make the changes.

* I think you have a good point re: the CRLF at the beginning of the boundary.  If that were included, I think that would eliminate the need for the ugly "-2" index adjustments.  As I mentioned before, the initial CRLF appears to not be included in the request body by the requests library, i.e. it's removed as part of header parsing.

  If I were to prime the initial buffer with "\r\n" to reflect that bit, that may be a smaller wart and would allow for the rest of the logic to be more clean.

* Re: regex vs. substring matching for the boundary line: this is because there's 2 types of boundary lines and I was trying to allow for one pattern to match both.  Specifically: ```<CRLF>--<boundary_str><CRLF>``` in the normal case, and ```<CRLF>--<boundary_str>--<CRLF>``` in the terminal case.

  Because the "--" suffix for terminal strings is stuck in the middle of the pattern prior to the final CRLF, it felt reasonable to use a regex for that so I could simply detect a complete boundary without having to use 2 distinct strings.  However, you also had a good point; I did nothing to escape the boundary value.  So clearly, something should change there.

* Re: FSM or no: I'm open to alternatives.  I'd like to perform some reworks for the other items noted here first, but if that is not sufficient (or if you strongly disagree), I'm open to discussing this.

* Re: how things get fed into the buffer: I combine the remainder of the last chunk with the new chunk to ease matching for substrings/regexes.  That's the simple reason.  The buffer is continually trimmed throughout the stages of this FSM, and chunk sizes should not be terribly large, thus that's why I felt that was appropriate.  I am very happy to discuss this further though.